toolchain/base/value_ids.h

```diff
@@ -41,20 +50,147 @@ class Real : public Printable<Real> {
+  // Tries to make an unsigned APInt into an embedded value in the ID, and if
+  // unable to do that returns the `Invalid` ID.
+  static auto TryMakeUnsignedValue(llvm::APInt value) -> IntId {
+    if (value.ule(MaxValue)) {
+      return IntId(value.getZExtValue());
+    }
+
+    return Invalid;
+  }
+
+  // Construct an ID from a raw 32-bit ID value.
+  static auto MakeRaw(int32_t raw_id) -> IntId { return IntId(raw_id); }
+
+  constexpr auto is_valid() const -> bool { return id_ != InvalidId; }
+  constexpr auto is_value() const -> bool { return id_ > ZeroIndexId; }
+  constexpr auto is_index() const -> bool { return id_ <= ZeroIndexId; }
+
+  auto AsValue() const -> int {
+    CARBON_DCHECK(is_value());
+    return id_;
+  }
+
+  constexpr auto AsIndex() const -> int {
+    CARBON_DCHECK(is_index());
+    return ZeroIndexId - id_;
+  }
+
+  constexpr auto AsRaw() const -> int32_t { return id_; }
+
   auto Print(llvm::raw_ostream& out) const -> void {
-    out << "int";
-    IdBase::Print(out);
+    out << "int [";
+    if (is_value()) {
+      out << "value: " << AsValue() << "]";
+    } else if (is_index()) {
+      out << "index: " << AsIndex() << "]";
+    } else {
+      CARBON_CHECK(!is_valid());
+      out << "invalid]";
+    }
+  }
+
+  friend constexpr auto operator==(IntId lhs, IntId rhs) -> bool {
+    return lhs.id_ == rhs.id_;
+  }
+  friend constexpr auto operator<=>(IntId lhs, IntId rhs)
+      -> std::strong_ordering {
+    return lhs.id_ <=> rhs.id_;
   }
+
+ private:
+  // We need all the values from maximum to minimum and a healthy range of
+  // indices to all fit within the token ID bits.
+  //
+  // We represent this as a signed TokenIdBits-bit 2s compliment integer. The
+  // sign extension from TokenIdBits to a register size can be folded into the
+  // shift used to extract from compressed bitfield storage.
+  //
+  // We then divide the smallest 1/4th of the space to indices, and the larger
+  // 3/4ths to embedded values. For 23-bits total this still gives us 2 million
+  // unique integers larger than the embedded ones, which would be difficult to
+  // fill without exceeding the number of tokens we can lex (8 million). For
+  // non-token based integers, the indices can continue downward to the 32-bit
+  // signed integer minimum.
+  //
+  // Note that the invalid ID can't be used with a token. This is OK as we
+  // expect invalid tokens to be *error* tokens and not need to represent an
+  // invalid integer.
+  static constexpr int TokenIdBitsShift = 32 - TokenIdBits;
+  static constexpr int32_t MaxValue =
+      std::numeric_limits<int32_t>::max() >> TokenIdBitsShift;
+  static constexpr int32_t ZeroIndexId = std::numeric_limits<int32_t>::min() >>
+                                         (TokenIdBitsShift + 1);
+  static constexpr int32_t MinValue = ZeroIndexId + 1;
+  static constexpr int32_t InvalidId = std::numeric_limits<int32_t>::min();
+  static constexpr int32_t InvalidIndex = ZeroIndexId - InvalidId;
```

Comment:
Sure, I've tried to improve the comment structure a bit.

What I ended up doing is lifting the big comment block up to the `IntId` class since its now in its own file, and added dedicated comments for each constant to try and explain the interesting parts of it.

Let me know if I missed anything you would find useful or any of the new comments aren't reading well.